sdk/storage/AzureStorageBlob/Source/DataStructures/StorageStreamUploader.swift

```diff
@@ -0,0 +1,540 @@
+        defer { handle.closeFile() }
+        return handle.readDataToEndOfFile()
+    }
+
+    /// Uploads the entire blob in a parallel fashion.
+    /// - Parameters:
+    ///   - group: An optional `DispatchGroup` to wait for the download to complete.
+    ///   - completion: A completion handler called when the download completes.
+    public func complete(inGroup group: DispatchGroup? = nil, then completion: @escaping () -> Void) throws {
+        guard !isComplete else {
+            let progress = BlobUploadProgress(bytes: self.progress, totalBytes: fileSize)
+            if let delegate = self.delegate {
+                delegate.uploader(self, didFinishWithProgress: progress)
+            } else {
+                completion()
+            }
+            return
+        }
+        let defaultGroup = DispatchGroup()
+        let dispatchGroup = group ?? defaultGroup
+        for _ in blockList {
+            dispatchGroup.enter()
+            next(inGroup: dispatchGroup) { _, _ in
+                if let delegate = self.delegate {
+                    let progress = BlobUploadProgress(bytes: self.progress, totalBytes: self.fileSize)
+                    delegate.uploader(self, didUpdateWithProgress: progress)
+                }
+            }
+        }
+        dispatchGroup.notify(queue: DispatchQueue.main) {
+            // Once all blocks are done, commit block list
+            self.commit { result, _ in
+                switch result {
+                case let .success(data):
+                    let progress = BlobUploadProgress(bytes: self.progress, totalBytes: self.fileSize)
+                    if let delegate = self.delegate {
+                        delegate.uploader(self, didFinishWithProgress: progress)
+                    } else {
+                        completion()
+                    }
+                case let .failure(error):
+                    if let delegate = self.delegate {
+                        delegate.uploader(self, didFailWithError: error)
+                    }
+                }
+            }
+        }
+    }
+
+    public func commit(
+        requestId: String? = nil,
+        transactionalContentMd5: Data? = nil,
+        transactionalContentCrc64: Data? = nil,
+        inGroup _: DispatchGroup? = nil,
+        then completion: @escaping HTTPResultHandler<BlobProperties>
+    ) {
+        // Construct URL
+        let urlTemplate = "/{container}/{blob}"
+        let pathParams = [
+            "container": containerName,
+            "blob": blobName
+        ]
+        let url = client.url(forTemplate: urlTemplate, withKwargs: pathParams)
+
+        // Construct parameters
+        var queryParams = [
+            ("comp", "blocklist")
+        ]
+        if let timeout = options.timeout { queryParams.append(("timeout", String(timeout))) }
+
+        // Construct headers
+        var headers = HTTPHeaders([
+            .contentType: "application/xml; charset=utf-8",
```

Comment:
Maybe we should pull the content-types (this one & "octet-stream") up into constants as well. Seems like they're well-defined values needed for the Storage APIs.

Reply:
So, I could create an enum with common ContentType values, in which case the call looks like this:

```Swift
var headers = HTTPHeaders([
  .accept: ContentType.xml.rawValue, // new
  .accept: "application/xml"         // old
])
```

I'm not sure the the enum adds value here due to the awkward calling mechanism and need to use rawValue. Also, I don't think we'd be able to come up with sensible names for each type, so we'd have to leave the raw string approach open.

Plain constants would work, but there's still the question of having names for each possible value, and they would be much less discoverable than an enum.

My preference here would be to keep the raw string.

Reply:
I don't feel strongly about it.